fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+			    .detail("AuditType", auditType)
+			    .detail("RetryCount", currentRetryCount)
+			    .detail("Contains", self->audits.contains(auditType) && self->audits[auditType].contains(auditID));
+			wait(delay(0.1));
+			TraceEvent(SevVerbose, "DDAuditStorageCoreRetryAfterWait", self->ddId)
+			    .detail("AuditID", auditID)
+			    .detail("AuditType", auditType)
+			    .detail("RetryCount", currentRetryCount)
+			    .detail("Contains", self->audits.contains(auditType) && self->audits[auditType].contains(auditID));
+			// Erase the old audit from map and spawn a new audit inherit from the old audit
+			removeAuditFromAuditMap(self,
+			                        audit->coreState.getType(),
+			                        audit->coreState.id,
+			                        audit->coreState.getPhase(),
+			                        currentRetryCount); // remove audit
+			runAuditStorage(self, audit->coreState, audit->retryCount, "auditStorageCoreRetry");
+		} else {
+			try {
+				audit->coreState.setPhase(AuditPhase::Failed);
+				wait(persistAuditState(self->txnProcessor->context(),
+				                       audit->coreState,
+				                       "AuditStorageCoreError",
+				                       lockInfo,
+				                       self->context->isDDEnabled()));
+				TraceEvent(SevInfo, "DDAuditStorageCoreSetFailed", self->ddId)
+				    .detail("Context", context)
+				    .detail("AuditID", auditID)
+				    .detail("AuditType", auditType)
+				    .detail("RetryCount", currentRetryCount)
+				    .detail("AuditState", audit->coreState.toString())
+				    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+				removeAuditFromAuditMap(self,
+				                        audit->coreState.getType(),
+				                        audit->coreState.id,
+				                        audit->coreState.getPhase(),
+				                        currentRetryCount); // remove audit
+			} catch (Error& e) {
+				TraceEvent(SevWarn, "DDAuditStorageCoreErrorWhenSetAuditFailed", self->ddId)
+				    .errorUnsuppressed(e)
+				    .detail("Context", context)
+				    .detail("AuditID", auditID)
+				    .detail("AuditType", auditType)
+				    .detail("RetryCount", currentRetryCount)
+				    .detail("AuditState", audit->coreState.toString())
+				    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+				// unexpected error when persistAuditState
+				// However, we do not want any audit error kills the DD
+				// So, we silently remove audit from auditMap
+				// As a result this audit can be in RUNNING state on disk but not alive
+				// We call this audit a zombie audit
+				// The client can wait on the state on disk changing to "complete"
+				// However, this change can never happen to a zombie audit
+				// For this case, the client should be able to be timed out
+				// A zombie aduit will be either: (1) resumed by the next DD; (2) removed by client
+				removeAuditFromAuditMap(self,
```

Comment:
nit: Move this to line 1782, and remove `removeAuditFromAuditMap` in the try block.